File: packages/aws-cdk-dotnet/src/Amazon.CDK.AWS.CertificateManager/CertificateRef.cs
using Amazon.CDK;
using AWS.Jsii.Runtime.Deputy;

namespace Amazon.CDK.AWS.CertificateManager
{
    /// <summary>Interface for certificate-like objects</summary>
    [JsiiClass(typeof(CertificateRef), "@aws-cdk/aws-certificatemanager.CertificateRef", "[{\"name\":\"parent\",\"type\":{\"fqn\":\"@aws-cdk/cdk.Construct\"}},{\"name\":\"name\",\"type\":{\"primitive\":\"string\"}}]")]
    public abstract class CertificateRef : Construct
    {
        protected CertificateRef(Construct parent, string name): base(new DeputyProps(new object[]{parent, name}))
        {
        }

        protected CertificateRef(ByRefValue reference): base(reference)
        {
        }

        protected CertificateRef(DeputyProps props): base(props)
        {
        }

        [JsiiProperty("certificateArn", "{\"fqn\":\"@aws-cdk/aws-certificatemanager.CertificateArn\"}")]
        public virtual CertificateArn CertificateArn
        {
            get => GetInstanceProperty<CertificateArn>();
        }

        /// <summary>Import a certificate</summary>
        [JsiiMethod("import", "{\"fqn\":\"@aws-cdk/aws-certificatemanager.CertificateRef\"}", "[{\"name\":\"parent\",\"type\":{\"fqn\":\"@aws-cdk/cdk.Construct\"}},{\"name\":\"name\",\"type\":{\"primitive\":\"string\"}},{\"name\":\"props\",\"type\":{\"fqn\":\"@aws-cdk/aws-certificatemanager.CertificateRefProps\"}}]")]
        public static CertificateRef Import(Construct parent, string name, ICertificateRefProps props)
        {
            return InvokeStaticMethod<CertificateRef>(typeof(CertificateRef), new object[]{parent, name, props});
        }

        /// <summary>Export this certificate from the stack</summary>
        [JsiiMethod("export", "{\"fqn\":\"@aws-cdk/aws-certificatemanager.CertificateRefProps\"}", "[]")]
        public virtual ICertificateRefProps Export()
        {
            return InvokeInstanceMethod<ICertificateRefProps>(new object[]{});
        }
    }
}

File: packages/@aws-cdk/aws-certificatemanager/lib/certificate-ref.ts
import { Construct, Output } from "@aws-cdk/cdk";

/**
 * Interface for certificate-like objects
 */
export abstract class CertificateRef extends Construct {
  /**
   * Import a certificate
   */
  public static import(parent: Construct, name: string, props: CertificateRefProps): CertificateRef {
    return new ImportedCertificate(parent, name, props);
  }

  public abstract readonly certificateArn: string;

  /**
   * Export this certificate from the stack
   */
  public export(): CertificateRefProps {
    return {
      certificateArn: new Output(this, 'Arn', { value: this.certificateArn }).makeImportValue().toString()
    };
  }
}

/**
 * A Certificate that has been imported from another stack
 */
class ImportedCertificate extends CertificateRef {
  public readonly certificateArn: string;

  constructor(parent: Construct, name: string, props: CertificateRefProps) {
    super(parent, name);

    this.certificateArn = props.certificateArn;
  }
}

/**
 * Reference to an existing Certificate
 */
export interface CertificateRefProps {
  /**
   * The certificate's ARN
   */
  certificateArn: string;
}
